File: backend/src/routes/ai.rs
//! AI 相关路由

use rocket::{State, serde::json::Json, http::Status};
use mongodb::Database;
use mongodb::bson::{doc, oid::ObjectId};
use std::str::FromStr;
use sha1::{Sha1, Digest};

use crate::models::{
    ApiResponse, Post, Note, Page,
    TimeCapsule, TimeCapsuleRequest, TimeCapsuleResponse, TimeSensitivity,
};
use crate::services::{AiService, ChatMessage, ChatRole};
use crate::db_find_one;

/// 内容信息结构
struct ContentInfo {
    title: String,
    text: String,
}

/// 计算内容 SHA1 哈希
fn compute_sha1(content: &str) -> String {
    let mut hasher = Sha1::new();
    hasher.update(content.as_bytes());
    format!("{:x}", hasher.finalize())
}

/// 构建 Time Capsule 分析的 prompt
fn build_analysis_prompt(title: &str, content: &str) -> Vec<ChatMessage> {
    let system_prompt = r###"你是一个内容时效性分析专家。你的任务是分析文章内容，判断其是否属于"易过期"类型。

易过期内容的特征包括：
- 特定版本的技术/框架/库（如 "React 18", "Node.js 20"）
- 具体的 API 接口或配置方式
- 时效性新闻或事件
- 特定日期的活动或促销
- 依赖外部服务的教程
- 软件安装/配置步骤
- 价格信息

不易过期内容的特征包括：
- 编程概念和原理
- 算法和数据结构
- 设计模式
- 人生感悟/随笔
- 历史事件回顾
- 数学/物理等基础知识

请以 JSON 格式返回分析结果：
{
  "sensitivity": "high" | "medium" | "low",
  "reason": "简短的分析理由（80字以内）",
  "markers": ["检测到的易过期元素1", "元素2", ...]
}

只返回 JSON，不要其他内容。"###;

    let user_prompt = format!(
        "请分析以下文章的时效性：\n\n标题：{title}\n\n内容：\n{content}"
    );

    vec![
        ChatMessage { role: ChatRole::System, content: system_prompt.to_string() },
        ChatMessage { role: ChatRole::User, content: user_prompt },
    ]
}

/// 解析 AI 返回的 JSON
fn parse_ai_response(response: &str) -> Result<(TimeSensitivity, String, Vec<String>), String> {
    eprintln!("AI raw response: {response}");

    // 尝试提取 JSON 部分
    let json_str = response
        .trim()
        .trim_start_matches("```json")
        .trim_start_matches("```")
        .trim_end_matches("```")
        .trim();

    // 尝试找到 JSON 对象的起始和结束位置
    let json_str = if let (Some(start), Some(end)) = (json_str.find('{'), json_str.rfind('}')) {
        &json_str[start..=end]
    } else {
        json_str
    };

    let parsed: serde_json::Value = serde_json::from_str(json_str)
        .map_err(|e| format!("Failed to parse AI response: {e}"))?;

    let sensitivity = match parsed.get("sensitivity").and_then(|v| v.as_str()).unwrap_or("medium") {
        "high" => TimeSensitivity::High,
        "low" => TimeSensitivity::Low,
        _ => TimeSensitivity::Medium,
    };

    let reason = parsed.get("reason")
        .and_then(|v| v.as_str())
        .unwrap_or("无法获取分析理由")
        .to_string();

    let markers = parsed.get("markers")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    Ok((sensitivity, reason, markers))
}

/// 分析文章时效性
#[utoipa::path(
    post,
    path = "/api/ai/time-capsule",
    request_body = TimeCapsuleRequest,
    responses(
        (status = 200, description = "成功分析文章时效性", body = ApiResponse<TimeCapsuleResponse>),
        (status = 400, description = "无效的请求参数"),
        (status = 404, description = "文章不存在"),
        (status = 500, description = "服务器内部错误"),
        (status = 503, description = "AI服务不可用")
    ),
    tag = "AI服务"
)]
#[post("/ai/time-capsule", data = "<request>")]
pub async fn analyze_time_capsule(
    db: &State<Database>,
    request: Json<TimeCapsuleRequest>,
) -> Result<Json<ApiResponse<TimeCapsuleResponse>>, Status> {
    // 1. 获取文章内容
    let content_info = fetch_content_by_type(db, &request).await?;

    // 2. 计算当前内容的 SHA1
    let content_hash = compute_sha1(&format!("{}{}", content_info.title, content_info.text));

    // 3. 检查缓存
    if let Some(cached) = get_cached_capsule(db, &request.ref_id, &content_hash).await {
        return Ok(Json(ApiResponse::success(TimeCapsuleResponse {
            sensitivity: cached.sensitivity,
            reason: cached.reason,
            markers: cached.markers,
            is_new: false,
        })));
    }

    // 4. 调用 AI 服务分析
    let (sensitivity, reason, markers) = analyze_with_ai(db, &content_info).await?;

    // 5. 保存到数据库
    save_time_capsule(db, &request, &content_hash, &sensitivity, &reason, &markers).await?;

    Ok(Json(ApiResponse::success(TimeCapsuleResponse {
        sensitivity,
        reason,
        markers,
        is_new: true,
    })))
}

/// 根据类型获取内容
async fn fetch_content_by_type(
    db: &State<Database>,
    request: &TimeCapsuleRequest,
) -> Result<ContentInfo, Status> {
    match request.ref_type.as_str() {
        "post" => {
            let object_id = ObjectId::from_str(&request.ref_id)
                .map_err(|_| Status::BadRequest)?;
            let collection = db.collection::<Post>("posts");
            let post = db_find_one!(collection, doc! { "_id": object_id })?;
            Ok(ContentInfo { title: post.title, text: post.text })
        }
        "note" => {
            let object_id = ObjectId::from_str(&request.ref_id)
                .map_err(|_| Status::BadRequest)?;
            let collection = db.collection::<Note>("notes");
            let note = db_find_one!(collection, doc! { "_id": object_id })?;
            Ok(ContentInfo { title: format!("日记 #{}", note.nid), text: note.text })
        }
        "page" => {
            let collection = db.collection::<Page>("pages");
            let page = db_find_one!(collection, doc! { "slug": &request.ref_id })?;
            Ok(ContentInfo { title: page.title, text: page.text })
        }
        _ => Err(Status::BadRequest),
    }
}

/// 获取缓存的时间胶囊
async fn get_cached_capsule(
    db: &State<Database>,
    ref_id: &str,
    content_hash: &str,
) -> Option<TimeCapsule> {
    let capsules_collection = db.collection::<TimeCapsule>("time_capsules");
    let find_options = mongodb::options::FindOneOptions::builder()
        .sort(doc! { "created": -1 })
        .build();

    eprintln!("Looking for refId: {ref_id}, current_hash: {content_hash}");

    match capsules_collection
        .find_one(doc! { "refId": ref_id })
        .with_options(find_options)
        .await
    {
        Ok(Some(existing)) => {
            eprintln!("Found existing record, db_hash: {}, current_hash: {}", existing.hash, content_hash);
            if existing.hash == content_hash {
                eprintln!("Hash matched, returning cached result");
                Some(existing)
            } else {
                eprintln!("Hash mismatch, will call AI");
                None
            }
        }
        Ok(None) => {
            eprintln!("No existing record found");
            None
        }
        Err(e) => {
            eprintln!("Database query error: {e:?}");
            None
        }
    }
}

/// 使用 AI 分析内容
async fn analyze_with_ai(
    db: &State<Database>,
    content_info: &ContentInfo,
) -> Result<(TimeSensitivity, String, Vec<String>), Status> {
    let ai_service = AiService::from_database(db.inner())
        .await
        .map_err(|e| {
            eprintln!("Failed to init AI service: {e}");
            Status::InternalServerError
        })?;

    if !ai_service.is_enabled() {
        return Err(Status::ServiceUnavailable);
    }

    let messages = build_analysis_prompt(&content_info.title, &content_info.text);
    let ai_response = ai_service
        .chat(messages, Some(0.3), None)
        .await
        .map_err(|e| {
            eprintln!("AI request failed: {e}");
            Status::InternalServerError
        })?;

    parse_ai_response(&ai_response).map_err(|e| {
        eprintln!("Failed to parse AI response: {e}");
        Status::InternalServerError
    })
}

/// 保存时间胶囊到数据库
async fn save_time_capsule(
    db: &State<Database>,
    request: &TimeCapsuleRequest,
    content_hash: &str,
    sensitivity: &TimeSensitivity,
    reason: &str,
    markers: &[String],
) -> Result<(), Status> {
    let capsules_collection = db.collection::<TimeCapsule>("time_capsules");
    let new_capsule = TimeCapsule {
        id: ObjectId::new(),
        ref_id: request.ref_id.clone(),
        ref_type: request.ref_type.clone(),
        sensitivity: sensitivity.clone(),
        reason: reason.to_string(),
        markers: markers.to_vec(),
        hash: content_hash.to_string(),
        created: bson::Bson::DateTime(bson::DateTime::now()),
    };

    capsules_collection
        .insert_one(&new_capsule)
        .await
        .map_err(|e| {
            eprintln!("Failed to save time capsule: {e}");
            Status::InternalServerError
        })?;

    Ok(())
}

/// 获取文章的 Time Capsule 分析结果
#[utoipa::path(
    get,
    path = "/api/ai/time-capsule/{ref_id}",
    params(
        ("ref_id" = String, Path, description = "文章ID")
    ),
    responses(
        (status = 200, description = "成功获取分析结果", body = ApiResponse<TimeCapsuleResponse>),
        (status = 404, description = "分析结果不存在"),
        (status = 500, description = "服务器内部错误")
    ),
    tag = "AI服务"
)]
#[get("/ai/time-capsule/<ref_id>")]
pub async fn get_time_capsule(
    db: &State<Database>,
    ref_id: &str,
) -> Result<Json<ApiResponse<TimeCapsuleResponse>>, Status> {
    let capsules_collection = db.collection::<TimeCapsule>("time_capsules");

    let find_options = mongodb::options::FindOneOptions::builder()
        .sort(doc! { "created": -1 })
        .build();

    let capsule = capsules_collection
        .find_one(doc! { "refId": ref_id })
        .with_options(find_options)
        .await
        .map_err(|_| Status::InternalServerError)?
        .ok_or(Status::NotFound)?;

    Ok(Json(ApiResponse::success(TimeCapsuleResponse {
        sensitivity: capsule.sensitivity,
        reason: capsule.reason,
        markers: capsule.markers,
        is_new: false,
    })))
}

File: backend_ts/src/routes/ai.ts
import type { TimeCapsuleAnalyzeRequest } from "@/types/api";
import type { TimeCapsule, TimeCapsuleResponse, TimeSensitivity } from "@/types/models";
import { Elysia, t } from "elysia";
import { ObjectId } from "mongodb";
import { getConfig } from "@/config";
import { db_find, db_insert, db_read } from "@/lib/db";
import { responsePlugin } from "@/plugins/response";

const config = getConfig();
const DB_NAME = config.mongodb.database || "mx-space";
const TIME_CAPSULES_COLLECTION = "time_capsules";

/**
 * 计算内容的 SHA1 哈希值
 */
async function computeSha1(content: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(content);
  const hashBuffer = await crypto.subtle.digest("SHA-1", data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map(b => b.toString(16).padStart(2, "0")).join("");
}

/**
 * 获取缓存的时间胶囊分析结果
 */
async function getCachedCapsule(
  refId: string,
  contentHash: string,
): Promise<TimeCapsule | null> {
  try {
    // 查询最新的分析结果
    const results = await db_read(
      DB_NAME,
      TIME_CAPSULES_COLLECTION,
      { refId },
      { sort: { created: -1 }, limit: 1 },
    );

    if (results.length === 0) {
      console.log(`[AI] No cached capsule found for refId: ${refId}`);
      return null;
    }

    const existing = results[0] as TimeCapsule;
    console.log(
      `[AI] Found cached capsule - db_hash: ${existing.hash}, current_hash: ${contentHash}`,
    );

    // 检查内容哈希是否匹配
    if (existing.hash === contentHash) {
      console.log("[AI] Hash matched, returning cached result");
      return existing;
    }

    console.log("[AI] Hash mismatch, need new analysis");
    return null;
  } catch (err) {
    console.error("[AI] Error fetching cached capsule:", err);
    return null;
  }
}

/**
 * 保存时间胶囊分析结果到数据库
 */
async function saveTimeCapsule(
  refId: string,
  refType: string,
  contentHash: string,
  sensitivity: TimeSensitivity,
  reason: string,
  markers: string[],
): Promise<boolean> {
  try {
    const capsule = {
      refId,
      refType,
      sensitivity,
      reason,
      markers,
      hash: contentHash,
      created: new Date(),
    };

    const success = await db_insert(DB_NAME, TIME_CAPSULES_COLLECTION, capsule);
    if (success) {
      console.log(`[AI] Saved time capsule for refId: ${refId}`);
    } else {
      console.error(`[AI] Failed to save time capsule for refId: ${refId}`);
    }
    return success;
  } catch (err) {
    console.error("[AI] Error saving time capsule:", err);
    return false;
  }
}

/**
 * 模拟 AI 分析（因为实际 AI 服务未实现）
 * 在实际环境中，这里应该调用真实的 AI 服务
 */
function mockAiAnalysis(
  title: string,
  content: string,
): { sensitivity: TimeSensitivity; reason: string; markers: string[] } {
  // 简单的关键词检测逻辑
  const text = `${title} ${content}`.toLowerCase();
  const markers: string[] = [];

  // 检测技术版本号
  const versionPatterns = [
    /react\s+\d+/gi,
    /node\.?js\s+\d+/gi,
    /vue\s+\d+/gi,
    /angular\s+\d+/gi,
    /python\s+\d+/gi,
  ];

  for (const pattern of versionPatterns) {
    const matches = text.match(pattern);
    if (matches) {
      markers.push(...matches);
    }
  }

  // 检测 API 相关
  if (text.includes("api") && (text.includes("endpoint") || text.includes("接口"))) {
    markers.push("API 接口");
  }

  // 检测配置相关
  if (text.includes("config") || text.includes("配置")) {
    markers.push("配置方式");
  }

  // 根据检测结果判断时效性
  let sensitivity: TimeSensitivity;
  let reason: string;

  if (markers.length >= 3) {
    sensitivity = "high";
    reason = "内容包含多个技术版本、API 或配置相关信息，属于高时效性内容";
  } else if (markers.length >= 1) {
    sensitivity = "medium";
    reason = "内容包含部分技术细节，可能随时间变化";
  } else {
    sensitivity = "low";
    reason = "内容主要为概念性或原理性知识，不易过期";
  }

  return { sensitivity, reason, markers };
}

/**
 * AI 路由处理器
 *
 * 提供以下端点：
 * - POST /ai/time-capsule/analyze - 分析时间胶囊
 * - GET /ai/time-capsule/:refId - 获取分析结果
 */
export const aiRoutes = new Elysia({ prefix: "/ai" })
  .use(responsePlugin())

  // POST /ai/time-capsule/analyze - 分析时间胶囊
  .post(
    "/time-capsule/analyze",
    async ({ body, success, error, set }) => {
      try {
        const { refId, refType, content } = body;

        // 1. 输入验证
        if (!refId || !refType || !content) {
          set.status = 400;
          return error(400, "Missing required fields: refId, refType, content");
        }

        if (!["post", "note"].includes(refType)) {
          set.status = 400;
          return error(400, "Invalid refType, must be 'post' or 'note'");
        }

        // 验证 refId 是否为有效的 ObjectId
        if (!ObjectId.isValid(refId)) {
          set.status = 400;
          return error(400, "Invalid refId format");
        }

        // 2. 获取内容并计算哈希
        const contentHash = await computeSha1(content);
        console.log(`[AI] Analyzing refId: ${refId}, hash: ${contentHash}`);

        // 3. 检查缓存
        const cached = await getCachedCapsule(refId, contentHash);
        if (cached) {
          const response: TimeCapsuleResponse = {
            sensitivity: cached.sensitivity,
            reason: cached.reason,
            markers: cached.markers,
            isNew: false,
          };
          return success(response);
        }

        // 4. 调用 AI 分析（这里使用模拟实现）
        // 注意：在实际环境中，这里应该调用真实的 AI 服务
        // 如果 AI 服务不可用，应该返回 503 错误
        console.log("[AI] No cache found, performing analysis");

        // 从 content 中提取标题（假设格式为 "标题\n\n内容"）
        const lines = content.split("\n");
        const title = lines[0] || "Untitled";
        const text = lines.slice(1).join("\n");

        const analysis = mockAiAnalysis(title, text);

        // 5. 保存分析结果
        const saved = await saveTimeCapsule(
          refId,
          refType,
          contentHash,
          analysis.sensitivity,
          analysis.reason,
          analysis.markers,
        );

        if (!saved) {
          console.error("[AI] Failed to save analysis result");
          // 即使保存失败，也返回分析结果
        }

        const response: TimeCapsuleResponse = {
          sensitivity: analysis.sensitivity,
          reason: analysis.reason,
          markers: analysis.markers,
          isNew: true,
        };

        return success(response);
      } catch (err) {
        console.error("[AI] Error analyzing time capsule:", err);
        set.status = 500;
        return error(500, "Failed to analyze time capsule");
      }
    },
    {
      body: t.Object({
        refId: t.String(),
        refType: t.Union([t.Literal("post"), t.Literal("note")]),
        content: t.String(),
      }),
    },
  )

  // GET /ai/time-capsule/:refId - 获取分析结果
  .get("/time-capsule/:refId", async ({ params, success, error, set }) => {
    try {
      const { refId } = params;

      // 验证 refId 格式
      if (!ObjectId.isValid(refId)) {
        set.status = 400;
        return error(400, "Invalid refId format");
      }

      // 查询最新的分析结果
      const results = await db_read(
        DB_NAME,
        TIME_CAPSULES_COLLECTION,
        { refId },
        { sort: { created: -1 }, limit: 1 },
      );

      if (results.length === 0) {
        set.status = 404;
        return error(404, "Time capsule analysis not found");
      }

      const capsule = results[0] as TimeCapsule;
      const response: TimeCapsuleResponse = {
        sensitivity: capsule.sensitivity,
        reason: capsule.reason,
        markers: capsule.markers,
        isNew: false,
      };

      return success(response);
    } catch (err) {
      console.error("[AI] Error fetching time capsule:", err);
      set.status = 500;
      return error(500, "Failed to fetch time capsule");
    }
  });
